File: src/secrets.ts
import { readFileSync, statSync, readdirSync } from "fs";
import { join, extname } from "path";

export interface SecretFinding {
  file: string;
  line: number;
  type: string;
  match: string;
  severity: "high" | "medium" | "low";
}

const SKIP_DIRS = new Set([
  "node_modules", ".git", "dist", "build", ".next", "coverage",
  "__pycache__", ".venv", "venv", ".tox",
]);

const SKIP_EXTS = new Set([
  ".png", ".jpg", ".jpeg", ".gif", ".webp", ".ico", ".svg",
  ".pdf", ".zip", ".tar", ".gz", ".woff", ".woff2", ".ttf",
  ".eot", ".mp4", ".mp3", ".lock",
]);

const SECRET_PATTERNS: Array<{ name: string; regex: RegExp; severity: "high" | "medium" | "low" }> = [
  // High severity — known provider key formats
  { name: "AWS Access Key", regex: /\bAKIA[0-9A-Z]{16}\b/, severity: "high" },
  { name: "AWS Secret Key", regex: /aws_secret_access_key\s*=\s*["']?[A-Za-z0-9/+=]{40}["']?/i, severity: "high" },
  { name: "GitHub Token", regex: /\bgh[pousr]_[A-Za-z0-9_]{36,255}\b/, severity: "high" },
  { name: "Anthropic API Key", regex: /\bsk-ant-[A-Za-z0-9\-_]{90,}\b/, severity: "high" },
  { name: "OpenAI API Key", regex: /\bsk-(?:proj-)?[A-Za-z0-9]{20,}\b/, severity: "high" },
  { name: "Stripe Secret Key", regex: /\bsk_live_[A-Za-z0-9]{24,}\b/, severity: "high" },
  { name: "Stripe Test Key", regex: /\bsk_test_[A-Za-z0-9]{24,}\b/, severity: "medium" },
  { name: "Slack Bot Token", regex: /\bxoxb-[0-9]{10,}-[0-9]{10,}-[A-Za-z0-9]{24,}\b/, severity: "high" },
  { name: "Slack Webhook", regex: /https:\/\/hooks\.slack\.com\/services\/T[A-Z0-9]+\/B[A-Z0-9]+\/[A-Za-z0-9]+/, severity: "high" },
  { name: "Twilio Auth Token", regex: /(?:twilio|auth_token)\s*[:=]\s*["']?[a-f0-9]{32}["']?/i, severity: "high" },
  { name: "Sendgrid API Key", regex: /\bSG\.[A-Za-z0-9_\-]{22}\.[A-Za-z0-9_\-]{43}\b/, severity: "high" },
  { name: "Private Key Block", regex: /-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----/, severity: "high" },

  // Medium severity — generic patterns
  { name: "Generic API Key", regex: /(?:api[_-]?key|apikey)\s*[:=]\s*["']?[A-Za-z0-9_\-]{20,}["']?/i, severity: "medium" },
  { name: "Generic Secret", regex: /(?:secret|password|passwd|pwd)\s*[:=]\s*["'][^"']{8,}["']/i, severity: "medium" },
  { name: "Bearer Token", regex: /Authorization\s*:\s*["']?Bearer\s+[A-Za-z0-9\-._~+/]+=*["']?/i, severity: "medium" },
  { name: "Database URL", regex: /(?:postgres|mysql|mongodb|redis):\/\/[^:\s]+:[^@\s]+@[^\s]+/, severity: "medium" },
  { name: "JWT Secret", regex: /jwt[_\-]?secret\s*[:=]\s*["'][^"']{16,}["']/i, severity: "medium" },
];

function scanFile(filePath: string): SecretFinding[] {
  const findings: SecretFinding[] = [];
  let content: string;

  try {
    content = readFileSync(filePath, "utf-8");
  } catch {
    return findings;
  }

  const lines = content.split("\n");
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    for (const pattern of SECRET_PATTERNS) {
      const match = pattern.regex.exec(line);
      if (match) {
        const raw = match[0];
        // Redact everything after the first 6 chars
        const redacted = raw.length > 10
          ? raw.slice(0, 6) + "..." + raw.slice(-4)
          : raw.slice(0, 3) + "***";
        findings.push({
          file: filePath,
          line: i + 1,
          type: pattern.name,
          match: redacted,
          severity: pattern.severity,
        });
        break; // One finding per line per file
      }
    }
  }

  return findings;
}

function walkDir(dir: string, findings: SecretFinding[]): void {
  let entries: string[];
  try {
    entries = readdirSync(dir);
  } catch {
    return;
  }

  for (const entry of entries) {
    if (SKIP_DIRS.has(entry)) continue;
    const full = join(dir, entry);
    let stat;
    try {
      stat = statSync(full);
    } catch {
      continue;
    }
    if (stat.isDirectory()) {
      walkDir(full, findings);
    } else if (!SKIP_EXTS.has(extname(entry).toLowerCase())) {
      findings.push(...scanFile(full));
    }
  }
}

export function scanSecrets(target: string): SecretFinding[] {
  const findings: SecretFinding[] = [];
  let stat;
  try {
    stat = statSync(target);
  } catch {
    throw new Error(`Path not found: ${target}`);
  }

  if (stat.isDirectory()) {
    walkDir(target, findings);
  } else {
    findings.push(...scanFile(target));
  }

  return findings;
}

File: secrets.go
package main

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

type SecretFinding struct {
	File     string
	Line     int
	Type     string
	Match    string
	Severity string
}

type secretPattern struct {
	name     string
	re       *regexp.Regexp
	severity string
}

var patterns = []secretPattern{
	{"AWS Access Key", regexp.MustCompile(`AKIA[0-9A-Z]{16}`), "high"},
	{"AWS Secret Key", regexp.MustCompile(`(?i)aws.{0,20}secret.{0,20}['\"][0-9a-zA-Z/+]{40}['\"]`), "high"},
	{"GitHub Token", regexp.MustCompile(`ghp_[0-9a-zA-Z]{36}`), "high"},
	{"GitHub OAuth", regexp.MustCompile(`gho_[0-9a-zA-Z]{36}`), "high"},
	{"GitHub App Token", regexp.MustCompile(`(ghu|ghs|ghr)_[0-9a-zA-Z]{36}`), "high"},
	{"Anthropic API Key", regexp.MustCompile(`sk-ant-[a-zA-Z0-9\-_]{93}`), "high"},
	{"OpenAI API Key", regexp.MustCompile(`sk-[a-zA-Z0-9]{48}`), "high"},
	{"Stripe Secret Key", regexp.MustCompile(`sk_live_[0-9a-zA-Z]{24}`), "high"},
	{"Stripe Publishable Key", regexp.MustCompile(`pk_live_[0-9a-zA-Z]{24}`), "medium"},
	{"Slack Token", regexp.MustCompile(`xox[baprs]-[0-9]{12}-[0-9]{12}-[0-9a-zA-Z]{24}`), "high"},
	{"Slack Webhook", regexp.MustCompile(`https://hooks\.slack\.com/services/[A-Z0-9]{9}/[A-Z0-9]{9}/[a-zA-Z0-9]{24}`), "high"},
	{"Twilio SID", regexp.MustCompile(`AC[a-f0-9]{32}`), "medium"},
	{"Twilio Token", regexp.MustCompile(`(?i)twilio.{0,20}['\"][a-f0-9]{32}['\"]`), "high"},
	{"SendGrid Key", regexp.MustCompile(`SG\.[a-zA-Z0-9_\-]{22}\.[a-zA-Z0-9_\-]{43}`), "high"},
	{"Private Key Block", regexp.MustCompile(`-----BEGIN (RSA|EC|DSA|OPENSSH) PRIVATE KEY-----`), "high"},
	{"Generic Private Key", regexp.MustCompile(`-----BEGIN PRIVATE KEY-----`), "high"},
	{"Database URL", regexp.MustCompile(`(postgres|mysql|mongodb|redis|amqp)://[^:]+:[^@]+@[^\s"']+`), "high"},
	{"JWT Secret", regexp.MustCompile(`(?i)(jwt.secret|jwt_secret|jwt-secret)\s*[=:]\s*['\"][^'"]{16,}['\"]`), "high"},
	{"Generic API Key", regexp.MustCompile(`(?i)(api[_-]?key|apikey)\s*[=:]\s*['\"][a-zA-Z0-9_\-]{20,}['\"]`), "medium"},
	{"Generic Password", regexp.MustCompile(`(?i)(password|passwd|pwd)\s*[=:]\s*['\"][^'"]{8,}['\"]`), "medium"},
	{"Generic Secret", regexp.MustCompile(`(?i)(secret|token)\s*[=:]\s*['\"][a-zA-Z0-9_\-]{20,}['\"]`), "medium"},
}

var skipDirs = map[string]bool{
	"node_modules": true, ".git": true, "dist": true, "build": true,
	".next": true, "vendor": true, "__pycache__": true, ".cache": true,
	"target": true, "out": true, ".turbo": true,
}

var skipExts = map[string]bool{
	".png": true, ".jpg": true, ".jpeg": true, ".gif": true, ".ico": true,
	".svg": true, ".woff": true, ".woff2": true, ".ttf": true, ".eot": true,
	".pdf": true, ".zip": true, ".tar": true, ".gz": true, ".lock": true,
	".sum": true, ".min.js": true,
}

func scanSecrets(root string) ([]SecretFinding, error) {
	info, err := os.Stat(root)
	if err != nil {
		return nil, fmt.Errorf("cannot stat %s: %w", root, err)
	}

	var findings []SecretFinding

	if !info.IsDir() {
		findings, err = scanFile(root)
		return findings, err
	}

	err = filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			if skipDirs[d.Name()] {
				return filepath.SkipDir
			}
			return nil
		}

		ext := strings.ToLower(filepath.Ext(path))
		if skipExts[ext] {
			return nil
		}

		found, _ := scanFile(path)
		findings = append(findings, found...)
		return nil
	})

	return findings, err
}

func scanFile(path string) ([]SecretFinding, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, nil
	}

	// Skip binary files
	if strings.Contains(string(data[:min(512, len(data))]), "\x00") {
		return nil, nil
	}

	lines := strings.Split(string(data), "\n")
	var findings []SecretFinding

	for lineNum, line := range lines {
		for _, p := range patterns {
			if match := p.re.FindString(line); match != "" {
				findings = append(findings, SecretFinding{
					File:     path,
					Line:     lineNum + 1,
					Type:     p.name,
					Match:    redact(match),
					Severity: p.severity,
				})
			}
		}
	}

	return findings, nil
}

func redact(s string) string {
	if len(s) <= 10 {
		return "***"
	}
	return s[:6] + "..." + s[len(s)-4:]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
